balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,658 @@
+	newEndpoints := []resolver.Endpoint{}
+
+	for _, ep := range endpoints {
+		addrs := []resolver.Address{}
+		for _, addr := range ep.Addresses {
+			if _, ok := seenAddrs.Get(addr); ok {
+				continue
+			}
+			addrs = append(addrs, addr)
+		}
+		if len(addrs) == 0 {
+			continue
+		}
+		newEndpoints = append(newEndpoints, resolver.Endpoint{
+			Addresses:  addrs,
+			Attributes: ep.Attributes,
+		})
+	}
+	return newEndpoints
+}
+
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) reconcileSubConns(newEndpoints []resolver.Endpoint) {
+	// Remove old subConns that were not in new address list.
+	oldAddrs := resolver.NewAddressMap()
+	for _, k := range b.subConns.Keys() {
+		oldAddrs.Set(k, true)
+	}
+
+	// Flatten the new endpoint addresses.
+	newAddrs := resolver.NewAddressMap()
+	for _, endpoint := range newEndpoints {
+		for _, addr := range endpoint.Addresses {
+			newAddrs.Set(addr, true)
+		}
+	}
+
+	// Shut them down and remove them.
+	for _, oldAddr := range oldAddrs.Keys() {
+		if _, ok := newAddrs.Get(oldAddr); ok {
+			continue
+		}
+		val, _ := b.subConns.Get(oldAddr)
+		val.(*scData).subConn.Shutdown()
+		b.subConns.Delete(oldAddr)
+	}
+}
+
+// shutdownRemaining shuts down remaining subConns. Called when a subConn
+// becomes ready, which means that all other subConn must be shutdown.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) shutdownRemaining(selected *scData) {
+	for _, v := range b.subConns.Values() {
+		sd := v.(*scData)
+		if sd.subConn != selected.subConn {
+			sd.subConn.Shutdown()
+		}
+	}
+	for _, k := range b.subConns.Keys() {
+		b.subConns.Delete(k)
+	}
+	b.subConns.Set(selected.addr, selected)
+}
+
+// requestConnection starts connecting on the subchannel corresponding to the
+// current address. If no subchannel exists, one is created. If the current
+// subchannel is in TransientFailure, a connection to the next address is
+// attempted.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) requestConnection() {
+	if !b.addressList.isValid() || b.state == connectivity.Shutdown {
+		return
+	}
+	curAddr := b.addressList.currentAddress()
+	sd, ok := b.subConns.Get(curAddr)
+	if !ok {
+		var err error
+		// We want to assign the new scData to sd from the outer scope, hence
+		// we can't use := below.
+		sd, err = b.newSCData(curAddr)
+		if err != nil {
+			// This should never happen, unless the clientConn is being shut
+			// down.
+			b.logger.Warningf("Failed to create a subConn for address %v: %v", curAddr.String(), err)
+			// The LB policy remains in TRANSIENT_FAILURE until a new resolver
+			// update is received.
+			b.state = connectivity.TransientFailure
+			b.addressList.reset()
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.TransientFailure,
+				Picker:            &picker{err: fmt.Errorf("failed to create a new subConn: %v", err)},
+			})
```

Comment:
Is this proper to do here, or can it be reworked so this returns an error which results in the state being updated or something?  I know we set the state immediately before calling this function at least once.  I would expect to get exactly one `UpdateState` call from an LB policy for a single call in.  I.e. to handle a call in, we typically update our internal state, then have one function that aggregates that state and calls `UpdateState` if needed.